apiserver/facades/agent/upgrader/unitupgrader.go

```diff
@@ -213,52 +211,67 @@ func (u *UnitUpgraderAPI) Tools(ctx context.Context, args params.Entities) (para
 	return result, nil
 }
 
-func (u *UnitUpgraderAPI) getAssignedMachine(ctx context.Context, tag names.Tag) (*state.Machine, error) {
+func (u *UnitUpgraderAPI) getAssignedMachine(ctx context.Context, tag names.Tag) (machine.Name, error) {
 	// Check that we really have a unit tag.
 	switch tag := tag.(type) {
 	case names.UnitTag:
 		unitName, err := coreunit.NewName(tag.Id())
 		if err != nil {
-			return nil, apiservererrors.ErrPerm
+			return "", apiservererrors.ErrPerm
 		}
-		id, err := u.applicationService.GetUnitMachineName(ctx, unitName)
+		machineName, err := u.applicationService.GetUnitMachineName(ctx, unitName)
 		if errors.Is(err, applicationerrors.UnitNotFound) {
-			return nil, apiservererrors.ServerError(coreerrors.NotFound)
+			return "", apiservererrors.ServerError(coreerrors.NotFound)
 		} else if err != nil {
-			return nil, err
+			return "", err
 		}
-		return u.st.Machine(id.String())
+		return machineName, nil
 	default:
-		return nil, apiservererrors.ErrPerm
+		return "", apiservererrors.ErrPerm
 	}
 }
 
 func (u *UnitUpgraderAPI) getMachineTools(ctx context.Context, tag names.Tag) params.ToolsResult {
 	var result params.ToolsResult
-	machine, err := u.getAssignedMachine(ctx, tag)
+	machineName, err := u.getAssignedMachine(ctx, tag)
 	if err != nil {
 		result.Error = apiservererrors.ServerError(err)
 		return result
 	}
-	machineTools, err := machine.AgentTools()
-	if err != nil {
+
+	machineTools, err := u.modelAgentService.GetMachineAgentBinaryMetadata(ctx, machineName)
+	if errors.Is(err, machineerrors.MachineNotFound) {
+		result.Error = apiservererrors.ServerError(coreerrors.NotFound)
+		return result
+	} else if err != nil {
 		result.Error = apiservererrors.ServerError(err)
 		return result
 	}
 	// We are okay returning the tools for just the one API server
 	// address since the unit agent won't try to download tools that
 	// are already present on the machine.
-	result.ToolsList = tools.List{machineTools}
+	vers, err := semversion.ParseBinary(machineTools.Version.String())
```

Comment:
Can we just construct the semversion.Binary please? The release might also need to be "ubuntu" or "linux".

Reply:
Done